### DISCUSSION TITLE
Customizing Passes with passkit-generator

### DISCUSSION THREAD

**Discussion by mis101247** [alexandercerutti/passkit-generator#214] [Q&A] [answered]
Hello,

I'm working on customizing a pass using passkit-generator and I've run into some issues. I'd really appreciate your help with the following:

Problem with Folder Model:
I have an existing mypass.pass file. When I load it using the Folder Model, I can't change any styles or colors. Is this normal? Or is there a way to modify styles when using the Folder Model?

I found https://github.com/alexandercerutti/passkit-generator/discussions/191, so I change to create pass by Buffer Model.

Issue with Buffer Model:
To get around the above limitation, I tried using the Buffer Model to create my pass. However, now I can't add personalization.json and personalizationLogo.png files.

```javascript
const pass = new PKPass(
        {
          'personalization.json': Buffer.from(
            personalizationJsonString,
            'utf-8',
          ),
          'personalizationLogo.png': fs.readFileSync(personalizationLogo),
          'pass.json': Buffer.from(passJsonString),
          'strip.png': fs.readFileSync(resizedStrip),
          'strip@2x.png': fs.readFileSync(resizedStrip2x),
          'logo.png': fs.readFileSync(resizedLogo),
          'logo@2x.png': fs.readFileSync(resizedLogo2x),
          'icon.png': fs.readFileSync(resizedLogo2x),
          'icon@2x.png': fs.readFileSync(resizedIcon),
        },
        {
          wwdr,
          signerCert,
          signerKey,
          signerKeyPassphrase,
        },
        {
          // keys to be added or overridden
          serialNumber: 'xxxxxxx',
        },
      );
```

My specific questions are:

How can I correctly add personalization.json and personalizationLogo.png when using the Buffer Model?
Is there a best practice that allows both style flexibility and personalization?
Am I doing something wrong? If so, how can I fix it?

Thank you very much for your time and help!

**Comment by alexandercerutti** [comment, marked as answer]
Hey @mis101247, thanks for using passkit-generator!

Today I cannot help you as I'm away. I'll likely be able to help you maybe this night or tomorrow, not sure yet.

In the meanwhile, I suggest you reading the documentation. There are some details and methods I think you'll need, like the usage of `addBuffer()`, which is called underneath when you create a pass instance.

Also, I suggest you looking at the logs, as passkit-generator may print some things if there is a validation error.

Last but not least, I'd like to ask you to edit your first message and add how you were trying to change the colors or other info, when using folder model.

Let me know if you are able to solve. I'll examine your question more deeply ASAP, thanks for understanding.

**Reply by alexandercerutti** [reply]
@mis101247 Hey, I found a bit of time to check your question. Thanks for waiting.

> I have an existing mypass.pass file. When I load it using the Folder Model, I can't change any styles or colors. Is this normal? Or is there a way to modify styles when using the Folder Model?

How did you try changing them? They can be changed just like you did with `serialNumber`.

> To get around the above limitation, I tried using the Buffer Model to create my pass. However, now I can't add personalization.json and personalizationLogo.png files.

What does that mean? You don't have them in the final bundle? Did you add NFC details?

Let me know

**Reply by mis101247** [reply]
Hello @alexandercerutti,

Thank you very much for your prompt response. I truly appreciate your time and assistance.

Regarding your questions:

> How did you try changing them? They can be changed just like you did with serialNumber.

Using the Folder Model, I discovered that it is indeed possible to change the colors. My initial inability to do so was due to using HEX color codes instead of the required RGB format.

Thank you I will choose this way for generating passes.


> What does that mean? You don't have them in the final bundle? Did you add NFC details?

Using the Buffer Model
I have included the NFC information in the passJsonString. However, when using the Buffer Model, I'm encountering difficulties adding the personalization.json to the PKPass. I've attempted the following methods:

```javascript
      const passJson = {
        formatVersion: 1,
        ...
        nfc: {
          message: serialNumber,
          encryptionPublicKey:
            'yyyy',
        },
      };

      const passJsonString = JSON.stringify(passJson);

```

add personalization.json when creating or after crating using add buffer

```javascript
new PKPass(
        {
          'personalization.json': Buffer.from(
            personalizationJsonString,
            'utf-8',
          ),
```
 
or by addBuffer
```javascript
pass.addBuffer(
        'personalization.json',
        Buffer.from(personalizationJsonString, 'utf-8'),
      );

      pass.addBuffer['personalizationLogo.png'] =
        fs.readFileSync(resizedStrip2x);
```

**Reply by alexandercerutti** [reply]
@mis101247 what issues were you encountering with Personalization?

**Reply by mis101247** [reply]
@alexandercerutti 
I couldn't add personalization.json when I using the Buffer Model.

**Reply by alexandercerutti** [reply]
What does it mean "cannot add"? You have any log, a crash or did you open the pkpass file as zip and verified the file wasn't added?

**Reply by alexandercerutti** [reply]
Btw, this will never work: `addBuffer` is a method, not a property.

```js
pass.addBuffer['personalizationLogo.png'] = ...
```

**Reply by mis101247** [reply]
I add the personalization.json and .png buffer through new PKPass(), but when I convert the generated pkpass file to a zip file and open it, I don't see the personalization.json or related image files.

**Reply by mis101247** [reply]
And I will try to using pass.addBuffer('personalizationLogo.png'), let you know the result.

Thank you for your package :)

**Reply by alexandercerutti** [reply]
If personalization files are invalid or nfc property is not set, they get trimmed away and you should see a log in the console. If you don't see anything, there could be a bug.

Let me know.
